### PR TITLE
add subscription type

### PR DESCRIPTION
**Pull request by filintod**
# Description

Add type to subscriptions info stored in compstore and then sent on metadata.

## Issue reference

Please reference the issue this PR will close: https://github.com/dapr/dapr/issues/7396

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/runtime/pubsub/subscription.go [hunk 1]
```diff
@@ -2,6 +2,26 @@ package pubsub
...
+type SubscriptionType int32
+
+const (
+	SubscriptionTypeUnknown = iota
+	SubscriptionTypeDeclarative
+	SubscriptionTypeProgrammatic
+	SubscriptionTypeStreaming
+)
+
+var subscriptionTypeName = map[int32]string{
```
Comment (ItalyPaleAle) [inline pkg/runtime/pubsub/subscription.go:14]: Can you make this a `switch` inside the `String()` method? This way we avoid allocating a map in the global scope

Comment (filintod) [reply pkg/runtime/pubsub/subscription.go]: actually changed it to add it to the proto of subscription because that is how it was used in the metadata in api/universal.

File: pkg/api/http/metadata.go [hunk 2]
```diff
@@ -93,13 +93,15 @@ func (a *api) onGetMetadata() http.HandlerFunc {
 
 				// Copy the subscriptions into a custom struct
 				if len(out.GetSubscriptions()) > 0 {
-					subs := make([]metadataResponsePubsubSubscription, len(out.GetSubscriptions()))
+					subs := make([]MetadataResponsePubsubSubscription, len(out.GetSubscriptions()))
```
Comment (filintod) [inline pkg/api/http/metadata.go:96]: had to put this as exportable to be able to use in the integration tests. as mentioned in the TODO below it would be better/cleaner to change these to use proto but then one of the tests was failing as the return value was returning .Rules.Rules that was not expected in the tests, not a big deal but not sure how people might be using this.

File: pkg/api/http/metadata.go [hunk 3]
```diff
@@ -175,12 +177,13 @@ type metadataActorRuntime struct {
...
 }
 
-type metadataResponsePubsubSubscription struct {
+type MetadataResponsePubsubSubscription struct {
 	PubsubName      string                                   `json:"pubsubname"`
 	Topic           string                                   `json:"topic"`
 	Metadata        map[string]string                        `json:"metadata,omitempty"`
 	Rules           []metadataResponsePubsubSubscriptionRule `json:"rules,omitempty"`
 	DeadLetterTopic string                                   `json:"deadLetterTopic"`
+	Type            runtimev1pb.PubsubSubscriptionType       `json:"type"`
```
Comment (filintod) [inline pkg/api/http/metadata.go:186]: this I'm not sure if use a string instead of the proto integer value

**Comment by filintod** [comment]
some format changes related to `make format`

File: pkg/api/http/metadata.go (hunk 2 above)
Comment (artursouza) [reply pkg/api/http/metadata.go:96]: Fair.

File: pkg/api/http/metadata.go (hunk 3 above)
Comment (artursouza) [reply pkg/api/http/metadata.go:186]: Please, use string. Otherwise, users of the HTTP API will need to consult a documentation to understand each number. For JSON, it is more intuitive to send the text instead of integer for enums.

**Review by artursouza — CHANGES_REQUESTED** [review]

**Comment by filintod** [comment]
sorry had to force push to deal with some bad dco merges

File: pkg/api/http/metadata.go (hunk 2 above)
Comment (JoshVanL) [reply pkg/api/http/metadata.go:96]: Please keep this type un-exported- where needed we copy the type across into the integration test as and where needed.

This API is not particularly "designed" out like the other first class APIs and exporting the types makes things worse. Long term, indeed these should be "promoted"  over to proto.

File: pkg/runtime/compstore/subscriptions.go [hunk 4]
```diff
@@ -187,6 +194,7 @@ func (c *ComponentStore) ListSubscriptionsAppByPubSub(name string) []*NamedSubsc
 			subs = append(subs, &NamedSubscription{Subscription: sub})
 		}
 	}
+	// TODO: there is no streaming here, is it ok??
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply pkg/runtime/compstore/subscriptions.go:197]: This func explicitly only returns non-streaming Subscriptions.

File: tests/e2e/bulk-pubsub/bulk_pubsub_test.go [hunk 5]
```diff
@@ -246,7 +246,8 @@ func testDropToDeadLetter(t *testing.T, publisherExternalURL, subscriberExternal
 			ReceivedByTopicCESub:      []string{},
 			ReceivedByTopicDead:       sentTopicDeadMessages,
 			ReceivedByTopicDeadLetter: sentTopicDeadMessages,
-			ReceivedByTopicCEBulkSub:  sentTopicCEMessages})
+			ReceivedByTopicCEBulkSub:  sentTopicCEMessages,
+		})
```
Comment (JoshVanL) [inline tests/e2e/bulk-pubsub/bulk_pubsub_test.go:250]: This file change doesn't look related to this PR, please revert.

File: tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/basic/http.go [hunk 6]
```diff
@@ -16,6 +16,12 @@ package basic
 import (
 	"context"
 	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	http2 "github.com/dapr/dapr/pkg/api/http"
+	rtv1 "github.com/dapr/dapr/pkg/proto/runtime/v1"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/basic/http.go:25]:
```suggestion
```

File: tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/bulk/http.go [hunk 7]
```diff
@@ -16,6 +16,13 @@ package bulk
...
 	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	http2 "github.com/dapr/dapr/pkg/api/http"
+	rtv1 "github.com/dapr/dapr/pkg/proto/runtime/v1"
 
 	"github.com/spiffe/go-spiffe/v2/spiffeid"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/bulk/http.go:27]: Please ensure imports are ordered.
```suggestion

	"github.com/spiffe/go-spiffe/v2/spiffeid"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	apihttp "github.com/dapr/dapr/pkg/api/http"
	rtv1 "github.com/dapr/dapr/pkg/proto/runtime/v1"
```

File: tests/integration/suite/daprd/subscriptions/programmatic/grpc/basic.go [hunk 8]
```diff
@@ -16,6 +16,9 @@ package grpc
 import (
 	"context"
 	"testing"
+	"time"
+
+	"github.com/dapr/dapr/pkg/api/http"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/programmatic/grpc/basic.go:21]: Same here - please order imports.

File: tests/integration/suite/daprd/subscriptions/programmatic/grpc/bulk.go [hunk 9]
```diff
@@ -16,11 +16,14 @@ package grpc
 import (
 	"context"
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"google.golang.org/protobuf/types/known/emptypb"
 
+	"github.com/dapr/dapr/pkg/api/http"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/programmatic/grpc/bulk.go:25]: Order imports.

File: tests/integration/suite/daprd/subscriptions/programmatic/http/v2alpha1/bulk.go [hunk 10]
```diff
@@ -16,6 +16,12 @@ package v2alpha1
 import (
 	"context"
 	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/programmatic/http/v2alpha1/bulk.go:21]: order imports.

File: tests/integration/suite/daprd/subscriptions/stream/basic.go [hunk 11]
```diff
@@ -18,6 +18,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/dapr/dapr/pkg/api/http"
+
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/stream/basic.go:22]: Order imports.

File: tests/integration/suite/daprd/subscriptions/stream/bulk.go [hunk 12]
```diff
@@ -18,6 +18,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/dapr/dapr/pkg/api/http"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/stream/bulk.go:21]: Order imports.

File: tests/integration/suite/daprd/subscriptions/stream/disconnect.go [hunk 13]
```diff
@@ -18,6 +18,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/dapr/dapr/pkg/api/http"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/stream/disconnect.go:21]: Order imports.

File: tests/integration/suite/daprd/subscriptions/stream/healthz.go [hunk 14]
```diff
@@ -18,6 +18,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/dapr/dapr/pkg/api/http"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/stream/healthz.go:21]: Order imports.

File: tests/integration/suite/daprd/subscriptions/stream/mixed.go [hunk 15]
```diff
@@ -18,6 +18,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/dapr/dapr/pkg/api/http"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/stream/mixed.go:21]: Order improts.

File: tests/integration/suite/daprd/subscriptions/stream/multi.go [hunk 16]
```diff
@@ -18,6 +18,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/dapr/dapr/pkg/api/http"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/stream/multi.go:21]: Order imports.

File: tests/perf/actor_double_activation/actor_double_activation_test.go [hunk 17]
```diff
@@ -34,9 +34,7 @@ const (
 	serviceApplicationName = "perf-actor-double-activation"
 )
 
-var (
```
Comment (JoshVanL) [inline tests/perf/actor_double_activation/actor_double_activation_test.go:37]: This file change doesn't look related to this PR, please revert.

File: tests/perf/actor_id_scale/actor_id_scale_test.go [hunk 18]
```diff
@@ -35,9 +35,7 @@ const (
 	actorType              = "scale-id"
 )
 
-var (
-	tr *runner.TestRunner
-)
+var tr *runner.TestRunner
```
Comment (JoshVanL) [inline tests/perf/actor_id_scale/actor_id_scale_test.go:38]: This file change doesn't look related to this PR, please revert.

File: tests/perf/configuration/configuration_test.go [hunk 19]
```diff
@@ -31,9 +31,7 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-var (
-	tr *runner.TestRunner
-)
+var tr *runner.TestRunner
```
Comment (JoshVanL) [inline tests/perf/configuration/configuration_test.go:34]: This file change doesn't look related to this PR, please revert.

File: tests/perf/pubsub_bulk_publish_http/pubsub_bulk_publish_http_test.go [hunk 20]
```diff
@@ -31,9 +31,7 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-var (
-	tr *runner.TestRunner
-)
+var tr *runner.TestRunner
```
Comment (JoshVanL) [inline tests/perf/pubsub_bulk_publish_http/pubsub_bulk_publish_http_test.go:34]: This file change doesn't look related to this PR, please revert.

File: tests/perf/pubsub_publish_http/pubsub_publish_http_test.go [hunk 21]
```diff
@@ -31,9 +31,7 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-var (
-	tr *runner.TestRunner
-)
+var tr *runner.TestRunner
```
Comment (JoshVanL) [inline tests/perf/pubsub_publish_http/pubsub_publish_http_test.go:34]: This file change doesn't look related to this PR, please revert.

File: tests/perf/pubsub_subscribe_http/pubsub_subscribe_test.go [hunk 22]
```diff
@@ -81,14 +81,14 @@ func getAppDescription(pubsubComponent Component, pubsubType string) kube.AppDes
 func TestMain(m *testing.M) {
 	utils.SetupLogs(testLabel)
 
-	//Read env variable for pubsub test config file, this will decide whether to run single component or all the components.
+	// Read env variable for pubsub test config file, this will decide whether to run single component or all the components.
```
Comment (JoshVanL) [inline tests/perf/pubsub_subscribe_http/pubsub_subscribe_test.go:84]: This file change doesn't look related to this PR, please revert.

File: tests/perf/utils/prometheus_metrics.go [hunk 23]
```diff
@@ -37,7 +37,6 @@ type DaprMetrics struct {
 
 // DAPR_PERF_METRICS_PROMETHEUS_PUSHGATEWAY_URL needs to be set
 func PushPrometheusMetrics(metrics DaprMetrics, perfTest, component string) {
-
```
Comment (JoshVanL) [inline tests/perf/utils/prometheus_metrics.go:40]: This file change doesn't look related to this PR, please revert.

File: tests/perf/workflows/workflow_test.go [hunk 24]
```diff
@@ -34,8 +34,10 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-var tr *runner.TestRunner
-var appNamePrefix = "perf-workflowsapp"
+var (
```
Comment (JoshVanL) [inline tests/perf/workflows/workflow_test.go:37]: This file change doesn't look related to this PR, please revert.

File: tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/bulk/http.go [hunk 25]
```diff
@@ -16,6 +16,12 @@ package bulk
...
 	"context"
 	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	rtv1 "github.com/dapr/dapr/pkg/proto/runtime/v1"
 
 	"github.com/spiffe/go-spiffe/v2/spiffeid"
```
Comment (JoshVanL) [inline tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/bulk/http.go:26]: Order imports.

**Review by JoshVanL — APPROVED** [review]

File: tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/bulk/http.go (hunk 25 above)
Comment (artursouza) [reply tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/bulk/http.go:26]: lint did not catch this.

Comment (filintod) [reply tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/bulk/http.go:26]: probably in `tests` it does not do it, similar to `make format` does not include tests

Comment (filintod) [reply tests/integration/suite/daprd/subscriptions/declarative/operator/v2alpha1/bulk/http.go:26]: I'm trying to finish conflict resolution on the buckets pr, and be back to this after

**Review by artursouza — APPROVED** [review]
